README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# leaflethex

<!-- badges: start -->

<!-- badges: end -->

The goal of leaflethex is to create hexbin layers that can be added to
plots made with the leaflet package.

This is a work in progress and is not yet a functional R package. Stay
tuned\!

## Installation

You can install the released version of leaflethex from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("leaflethex")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("rpruim/leaflethex")
```

## Example

This is a basic example which shows you how to add the hexbin layer to
your leaflet map:

``` r
library(leaflethex)
df = tibble(
  lat = rnorm(100),
  lng = rnorm(100)
)
map <- leaflet::leaflet(df) %>% 
  leaflet::addTiles() %>% 
  leaflethex::add_hexbin(
    radius = 19, 
    lowEndColor = "yellow", 
    highEndColor="red")
map
```

## Parameters for add\_hexbin

  - `data` - data frame or tibble - alternate data to use for this
    hexbin instead of default map data

  - `radius` - numeric - choose the base size for the hexagons

  - `opacity` - decimal - between 0.0 and 1.0 - choose the percent of
    opacity for the hexagons

  - `duration` - numeric - positive milliseconds taken by the renderer
    to draw the hexagons

  - `lowEndColor` - string - choose the color for the smaller hexagons

  - `highEndColor` - string - choose the color for the larger hexagons
    
    *Note About Coloring* - For best readable results, use colors that
    have high contrast in hue and shade such as red and yellow. White is
    generally a safe option for the `lowEndColor`, but the eye is better
    at picking up changes in hue than changes in shade.

  - `uniformSize` - boolean - choose if the size of the hexagons should
    be uniform or variable

  - `uniformColor` - string - choose a color to be uniform across the
    different sizes of hexagons instead of variable color

  - `sizeSummaryFunction` - string - choose a function from ‘count’,
    ‘sum’, ‘max’, ‘min’, ‘mean’, or ‘median’ or give a valid JS
    function (in development)

  - `sizevar` - string - the name of the variable used to calculate the
    size of the hex using the `sizeSummaryFunction`

  - `colorSummaryFunction` - string - choose a function from ‘count’,
    ‘sum’, ‘max’, ‘min’, ‘mean’, or ‘median’ or give a valid JS
    function (in development)

  - `colorvar` - string - the name of the variable used to calculate the
    color of the hex using the `colorSummaryFunction`


NEWS.md

# NEWS

## leaflethex 0.1.0 - First Release to CRAN

* Initial Upload

## leaflethex 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/hexbin.R
#' Hexbin Layers for Leaflet Plots
#'
#' Create hexbin layers for leaflet plots.
#' @importFrom dplyr %>%
#' @import leaflet
#' @param map The leaflet map object to apply the hexbin layer to.
#'   Makes this function compatible with the `%>%` operator
#' @param data data frame or tibble - alternate data to use for this hexbin
#'   instead of default map data
#' @param radius choose the base size for the hexagons
#' @param opacity decimal between 0.0 and 1.0 - choose the
#'   percent of opacity for the hexagons
#' @param duration positive integer milliseconds that the animation
#'   takes for drawing the hexagons
#' @param lowEndColor choose the color for the smaller hexagons
#' @param highEndColor choose the color for the larger hexagons
#' @param uniformSize a logical indicating whether all hexagons should be the same size.
#' @param uniformColor a color that overrides lowEndColor and highEndColor to
#'   make the color uniform across the hexagon sizes.
#' @param sizeSummaryFunction a string that specifies which summary function
#'   to use on sizevar to modulate the size of the hexagons.
#'   The options are 'count', 'sum', 'max', 'min', 'mean', and 'median'.
#' @param sizevar a string that specifies which variable in the user specified
#'   data frame will be used to calculate the size of the hexagons.
#' @param colorSummaryFunction a string that specifies which summary function
#'   to use on colorvar to modulate the color of the hexagons.
#'   The options are 'count', 'sum', 'max', 'min', 'mean', and 'median'.
#' @param colorvar a string that specifies which variable in the user specified
#'   data frame will be used to calculate the color of the hexagons.
#' @note Do not use uniformColor and uniformSize together as it will not give any
#'   insights to the data
#' @seealso A [JSFiddler Hexbin example](https://jsfiddle.net/reblace/acjnbu8t/?utm_source=website&utm_medium=embed&utm_campaign=acjnbu8t) by Ryan
#' @return map parameter, but with the hexbinLayer attached so that it can be
#'   used with the `%>%` pipe operator
#' @note If colorSummaryFunction and colorvar are not specified,
#'   the color will mirror the sizevar unless uniform color set to `TRUE`.
#' @examples
#' leaflet::leaflet(data.frame(lat =  42.9634 + rnorm(1000),lng = -85.6681 + rnorm(1000))) %>%
#' addTiles() %>% addHexbin()
#'
#' leaflet::leaflet(data.frame(lat =  42.9634 + rnorm(1000),lng = -85.6681 + rnorm(1000))) %>%
#' addTiles() %>% addHexbin(radius=25, lowEndColor='purple', highEndColor='orange')
#' @export
#'
#'
addHexbin <-
  function(map,
           data = NULL,
           radius = 12,
           opacity = 0.5,
           duration = 500,
           lowEndColor = "white",
           highEndColor = "blue",
           uniformSize = FALSE,
           uniformColor = NULL,
           sizeSummaryFunction = c("count", "sum", "max", "min", "mean", "median"),
           sizevar = NULL,
           colorSummaryFunction = c("count", "sum", "max", "min", "mean", "median"),
           colorvar = NULL) {

    sizeSummaryFunction  <- match.arg(sizeSummaryFunction)
    colorSummaryFunction <- match.arg(colorSummaryFunction)
    # Build MapData from given data or mapData if none provided
    mapData <- if(!is.null(data)) data else leaflet::getMapData(map)
    if(uniformSize && !is.null(uniformColor)) {
      warning("Using uniformSize and uniformColor together will not provide any insights into the data")
    }

    # Create the Hexbin Plugin
    addHex <-
      pluginFactory(
        "Hexbin",
        system.file("js", "", package = "leaflethex"),
        "hexbin.js", "deps.js", "hexbin.css")

    # Display warning if variables unset and function != "count"
    if(is.null(sizevar) && sizeSummaryFunction != "count") {
      line1 <- "`sizevar' must be set if using a summary function other than \"count\""
      line2 <- "The hexbin will calculate a simple count of data points per hex."
      warning(paste(line1, line2))
      sizeSummaryFunction  <- "count"
    }
    # Display warning if variables unset and function != "count"
    if(is.null(colorvar) && colorSummaryFunction != "count") {
      line1 <- "`colorvar' must be set if using a summary function other than \"count\""
      line2 <- "The hexbin will calculate a simple count of data points per hex."
      warning(paste(line1, line2))
      colorSummaryFunction  <- "count"
    }

    # Throw error if the variable chosen is not present in the data frame
    if(!is.null(sizevar) && ! sizevar %in% colnames(mapData)) {
      stop("The specified 'sizevar' being used is not in the given data frame. Perhaps there was a typo. 'sizevar' must be a string of the variable name")
    }
    if(!is.null(colorvar) && ! colorvar %in% colnames(mapData)) {
      stop("The specified 'colorvar' being used is not in the given data frame. Perhaps there was a typo. 'colorvar' must be a string of the variable name")
    }

    # If only one is set make the other one mirror (overridden by uniformColor/uniformSize)
    if(is.null(sizevar) && !is.null(colorvar)) {
      sizeSummaryFunction <- colorSummaryFunction
    }
    if(!is.null(sizevar) && is.null(colorvar)) {
      colorSummaryFunction <- sizeSummaryFunction
    }
    # Ensure the data passed to the JS script is a JSON object
    class(mapData) <- "options"
    # Pipe the Hexbin into the map
    map %>% addHex(data = mapData,
                   radius = radius,
                   opacity = opacity,
                   duration = duration,
                   lowEndColor = lowEndColor,
                   highEndColor = highEndColor,
                   uniformSize = uniformSize,
                   uniformColor = uniformColor,
                   sizeSummaryFunction = sizeSummaryFunction,
                   sizevar = sizevar,
                   colorSummaryFunction = colorSummaryFunction,
                   colorvar = colorvar)
  }

R/addJSPlugin.R
#' Create plugin functions to implement Leaflet pure-JS libraries in R
#'
#' @description `pluginFactory()` takes in a couple js files (code and dependencies)
#' and creates a plugin function that applies this plugin to a given map. The new
#' function also returns the map so it can be piped with the %>% operator like other
#' leaflet functions
#'
#' @import leaflet htmltools htmlwidgets tibble
#' @param name A name for your plugin
#' @param location The parent folder of your js file, dependency file, and css stylesheet
#' @param jsfilename A .js filename relative to the parent folder at `location` that holds the source code for modifying the map object
#' @param dependencies A .js filename(s) of all other js libraries the source code depends on
#' @param stylesheet A .css stylesheet for adding styles to the entire page
#' @return A function with arguments map, data, and ... to be used as a plugin to leaflet just like the leaflet::addCircles() function
#' @seealso Some code is used from [this github gist by Joe Chang](https://gist.github.com/jcheng5/c084a59717f18e947a17955007dc5f92)
#' @examples
#'
#' df <- data.frame(
#'   lat =  42.9634 + stats::rnorm(100),
#'   lng = -85.6681 + stats::rnorm(100),
#'   size = runif(100, 5, 20),
#'   color = sample(colors(), 100)
#' )
#' addJS <-
#'   pluginFactory(
#'     "Some JS Plugin",
#'     system.file("js", "", package = "leaflethex"), "hexbin.js", "deps.js", stylesheet="hexbin.css")
#'
#' leaflet(df, width = 600, height = 300) %>%
#'   addTiles() %>%
#'   addJS(radius = 20, highEndColor = "yellow")
#'
#' @export
pluginFactory <-
  function(
    name = "JSPlugin",
    location,
    jsfilename,
    dependencies,
    stylesheet = NULL) {

  # Create a HtmlDependency with the given files
  plugin <- createPlugin(name,
                         "1.0.0",
                         src=location,
                         script=dependencies,
                         stylesheet = stylesheet)
  # Create a function for the user to use as a map pipe
  function(map, data=NULL, ...){

    # Collect extra arguments
    dots <- list(...)

    # Get Default data if necessary
    mapData <- if(is.null(data)) leaflet::getMapData(map) else data

    # Attach mapData to the the parameters to be passed to the javascript
    dots$mapData = mapData

    # Ensure that the dots list and data is passed to the JS script is a JSON object
    class(dots) <- "options"

    # Read in JS Code
    jscode <- readJSFileToString(location, jsfilename)

    # Load Dependencies and render JS Code with the data and extra parameters
    registerPlugin(map, plugin) %>% htmlwidgets::onRender(jscode, data=dots)
  }
}

# functional wrapper for htmlDependency constructor
createPlugin <- function(name, version, src, script, stylesheet = NULL) {
  htmltools::htmlDependency(name, version, src = src, script = script, stylesheet = stylesheet)
}

# A function that takes a plugin htmlDependency object and adds
# it to the map. This ensures that however or whenever the map
# gets rendered, the plugin will be loaded into the browser.
registerPlugin <- function(map, plugin) {
  map$dependencies <- c(map$dependencies, list(plugin))
  map
}

# Extra utility function to convert .js files to strings
readJSFileToString <- function(location, jsfilename) {
  readr::read_file(paste(location, jsfilename, sep="/"))
}


inst/js/hexbin.js
function(el, x, data = null) {

  // Built-in Summary Functions
  var summaryFunctions = ["count", "sum", "max", "min", "mean", "median"];

  // Utility function for saving defaults for data parameters
  var param = function(name, defaultValue) { return { name: name, defaultValue : defaultValue}  };

  // Create a list of accepted parameters to this script
  var parameters = [
    param("radius", 12),
    param("opacity", 0.5),
    param("duration", 500),
    param("lowEndColor", 'white'),
    param("highEndColor", 'blue'),
    param("uniformSize", false),
    param("uniformColor", 'blue'),
    param("sizeSummaryFunction", "count"),
    param("colorSummaryFunction", "count"),
    param("sizevar", undefined),
    param("colorvar", undefined)
  ];

  // Function to build the options object for later use
  var buildOptions = function() {

    // Initialize the object that will be filled and returned
    var result = {};

    // Create options object with given values or default values if none are provided
    parameters.forEach(entry => {
      var parameterValue = data[entry.name];
      if(parameterValue !== undefined && parameterValue !== null) {
        result[entry.name] = parameterValue;
      } else {
        result[entry.name] = entry.defaultValue;
      }
    });
    // if no colorvar is set, then mirror with sizevar
    if (result.colorvar === undefined) {
      result.colorvar = result.sizevar;
    }
    // Return options chosen by the user plus defaults for unchosen fields
    return result;
  };
  // Set Options based on parameters given by the data parameters
  var options = buildOptions();


  // Builds a function to calculate per hexagon from the function name
  // or valid js string and variable name to use
  var buildSumFunction = function(functionToUse, variableToUse, variableType) {
    var retFunction;
    var count = dataPointsPerHex => dataPointsPerHex.length;
    // Default to using count
    if(functionToUse == "count" || variableToUse === undefined) return count;
    if(!summaryFunctions.includes(functionToUse)) {
      console.warn("Using Custom Function as no supported functions were matched");
      // User defined function
      retFunction = Function('"use strict";return (' + functionToUse + ')');
    } else {
      retFunction = function(dataPointsPerHex) {
        // Set the inital value for the result
        var result = 0;
        if(functionToUse == "max") result = Number.MIN_VALUE;
        if(functionToUse == "min") result = Number.MAX_VALUE;
        if(functionToUse == "sum") result = 0;

        console.log(dataPointsPerHex.length);
        if(functionToUse == "median") {
          var values = dataPointsPerHex.map(point => point.o[2][variableToUse]);
          if(values.length < 3) {
            if(values.length == 1) result = values[0];
            else if(values.length == 0) result = 0;
            else result = (values[0] + values[1]) / 2;
          } else {
            values.sort((a, b) => a - b);
            var index = (values.length + 1) / 2;
            if(Number.isInteger(index)) {
              result = values[index];
            } else {
              result = (values[Math.floor(index)] + values[Math.ceil(index)]) / 2;
            }
          }

        } else {
          dataPointsPerHex.forEach(point => {
            var value = point.o[2][variableToUse];
            if(functionToUse == "mean") result += value;
            if(functionToUse == "sum") result += value;
            if(functionToUse == "max") result = result > value ? result : value;
            if(functionToUse == "min") result = result < value ? result : value;
          });
          if(functionToUse == "mean") result /= dataPointsPerHex.length;
          if(variableType == "size") result = scaleFunctions[variableToUse](result);
        }
        return result;
      };
      return retFunction;
    }
  };

  // Add Options to hexlayer
  var hexLayer = L.hexbinLayer(options).addTo(this);

  // Add ColorScale
  var colorScale = [options.lowEndColor, options.highEndColor];
  if(data.uniformColor !== null && data.uniformColor !== undefined) {
    colorScale = [data.uniformColor, data.uniformColor];
  }
  hexLayer.colorScale().range(colorScale);

  // Set Radius range based on radius given through data parameter
  var largestRadius = options.radius - 1 / options.radius; // 1/radius is a small buffer between hexagons
  var smallestRadius = options.radius / 2;
  if(options.uniformSize) smallestRadius = largestRadius;
  var radRange = [ smallestRadius, largestRadius ];

  // Add Data to the Hex Layer
  var hexData = [];

  // Check if variable is an array with the same length as the latitute and longitude arrays
  var variableIsCompatible = entry => Array.isArray(entry) && entry.length == data.mapData.lat.length;
  var scaleFunctions = {};
  // Find any variables that may be involved in the hexbin
  var auxData = Object.assign({}, data.mapData);
  Object.keys(auxData).forEach(key => {
    if(key == "lat" || key == "lng") {
      delete auxData[key];
    } else if(variableIsCompatible(auxData[key])) {
      var array = auxData[key];
      var min = Math.min(...array);
      var max = Math.max(...array);
      var radSizeRange = (largestRadius - smallestRadius);
      var variableRange = (max - min);
      scaleFunctions[key] = function(variableValue) {
        return (((variableValue - min) * radSizeRange) / variableRange) + smallestRadius;
      };
    } else {
      delete auxData[key];
    }
  });

  for(i=0; i< data.mapData.lat.length; i++){
    var tuple = [data.mapData.lat[i], data.mapData.lng[i]];
    hexData.push(tuple);
  }
  if(auxData !== undefined) {
    for(i=0; i < hexData.length; i++) {
      var extraVarContainer = {};
      if(options.sizevar !== undefined) {
        extraVarContainer[options.sizevar] = auxData[options.sizevar][i];
      }
      if(options.colorvar !== undefined) {
        extraVarContainer[options.colorvar] = auxData[options.colorvar][i];
      }
      hexData[i].push(extraVarContainer);
    }
  }
  // hexData is now an Array of arrays (Matrix)
  // 0: [lat0, lng0, somevariable0, anothervariable0]
  // 1: [lat1, lng1, somevariable1, anothervariable1]
  // ...

  // Create Hex Layer
  hexLayer
  .radiusRange(radRange)
  .lat(function(d) { return d[0]; })
  .lng(function(d) { return d[1]; })
  .colorValue(buildSumFunction(options.colorSummaryFunction, options.colorvar, "color"))
  .radiusValue(buildSumFunction(options.sizeSummaryFunction, options.sizevar, "size")) // Choose summary function based off of parameter inside the data object
  .data(hexData);


  // Zoom in to fit the data in the screen
  this.fitBounds(hexData);

}